.doom.d/config.org
#+TITLE: doom emacs config
#+AUTHOR: flothepo

* Basic Configuration

** Doom
*** Theme and font
#+begin_src emacs-lisp :tangle yes
(setq doom-theme 'doom-vibrant
      doom-font (font-spec :family "Fira Code" :size 16)
      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 12)
      doom-big-font-mode nil
      doom-big-font-increment 2
      doom-themes-treemacs-theme "doom-colors")

(defvar my/themes
  (let ((themes '(doom-palenight
                  doom-opera-light
                  doom-vibrant)))
    (nconc themes themes)
    themes))

(defun my/cycle-theme ()
  (interactive)
  (setq doom-theme (car my/themes)
        my/themes (cdr my/themes))
  (doom/reload-theme)
  (message "Loaded theme %s" (symbol-name doom-theme)))
#+end_src

Bold keywords
#+begin_src emacs-lisp :tangle yes
(defun my-theme-customizations ()
 (set-face-attribute 'font-lock-keyword-face nil :weight 'bold))

(add-hook! 'doom-load-theme-hook #'my-theme-customizations)
#+end_src
Show Unicode glyphs using all-the-icons font
#+begin_src emacs-lisp :tangle yes
(set-fontset-font t 'unicode "all-the-icons" nil 'prepend)
#+end_src
**** Fira Code
Enable ligatures
#+begin_src emacs-lisp :tangle yes
(global-fira-code-mode)
(setq fira-code-mode-disabled-ligatures '("x" "[]"))
#+end_src
**** pretty symbols
unprettify prettified symbols when under cursor
#+begin_src emacs-lisp :tangle yes
(setq prettify-symbols-unprettify-at-point 'right-edge)
#+end_src
**** modeline
#+begin_src emacs-lisp :tangle yes
(setq doom-modeline-major-mode-icon t)
#+end_src
*** Misc
#+begin_src emacs-lisp :tangle yes
(setq custom-file (expand-file-name ".custom.el" doom-private-dir)
      doom-fallback-buffer-name "► Doom")
#+end_src
Ask which buffer to open after a window split
#+begin_src emacs-lisp :tangle yes
(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (+ivy/switch-buffer))
#+end_src
** Defaults
#+begin_src emacs-lisp :tangle yes
(setq-default
 tab-width 4
 uniquify-buffer-name-style 'forward
 window-combination-resize t
 x-stretch-cursor t)
#+end_src
#+begin_src emacs-lisp :tangle yes
(setq
 auto-save-default t
 inhibit-compacting-font-caches t)
#+end_src

** Basic modes
*** calendar
#+begin_src emacs-lisp :tangle yes
(setq calendar-week-start-day 1
      display-time-24hr-format t)
#+end_src
*** company
#+begin_src emacs-lisp :tangle yes
(setq company-tooltip-idle-delay 0.2
      company-idle-delay 0.2
      company-show-numbers t)
#+end_src
*** flyspell
Use hunspell
#+begin_src emacs-lisp :tangle yes
(setq ispell-program-name (executable-find "hunspell")
      ispell-dictionary "german")
(add-hook! org-mode #'flyspell-mode)
#+end_src

*** ivy
better matching
#+begin_src emacs-lisp :tangle yes
(setq ivy-re-builders-alist
      '((t . ivy--regex-plus)))
#+end_src
*** org
Visual stuff
#+begin_src emacs-lisp :tangle yes
(after! org
  (setq org-pretty-entities nil
        org-ellipsis  "▼"))

(add-hook! org-mode #'org-superstar-mode)
(setq org-superstar-prettify-item-bullets nil)
#+end_src
*** recentf
#+begin_src emacs-lisp :tangle yes
(after! recentf
  (add-to-list 'recentf-exclude
               (expand-file-name "~/.emacs.d//")))
#+end_src
*** treemacs
Do not show files ignored by git
#+begin_src emacs-lisp :tangle yes
(after! treemacs
  (add-to-list 'treemacs-pre-file-insert-predicates #'treemacs-is-file-git-ignored?))
#+end_src
** Language Configuration
*** Haskell
Use stack everywhere
#+begin_src emacs-lisp :tangle yes
(setq haskell-process-type 'stack-ghci
      haskell-process-path-stack "stack"
      haskell-process-args-stack-ghci '("--ghci-options=-ferror-spans"))
#+end_src
No annoying popups
#+begin_src emacs-lisp :tangle yes
(setq haskell-interactive-popup-errors nil)
#+end_src
**** hoogle
Some utility functions for interacting with hoogle
#+begin_src emacs-lisp :tangle yes
(defun hoogle-word-under-caret ()
  "Lookup the word under the caret on hoogle"
  (interactive)
  (save-excursion
    (forward-word)
    (let ((end (point)))
      (backward-word)
      (hoogle-region (point) end))))

(defun hoogle-region (start end)
  "Lookup the region between start and end on hoogle"
  (interactive "r")
  (haskell-hoogle (buffer-substring start end)))
#+end_src
** Keybindings
*** General
Make it easy to raise popups to be proper windows
#+begin_src emacs-lisp :tangle yes
(map! :leader
      :desc "Raise popup"
      "w m p" #'+popup/raise)
#+end_src
Wrap in parens
#+begin_src emacs-lisp :tangle yes
(map! :leader
      (:prefix "i"
       :desc "Wrap round"  "(" #'sp-wrap-round
       :desc "Wrap curly"  "{" #'sp-wrap-curly
       :desc "Wrap square" "[" #'sp-wrap-square))
#+end_src
[[flyspell]] in ivy
#+begin_src emacs-lisp :tangle yes
(map! :map flyspell-mode-map
      "C-;" #'flyspell-correct-wrapper)
#+end_src
Cycle doom themes
#+begin_src emacs-lisp :tangle yes
(map! :leader
      :desc "Cycle doom theme"
      "t t" #'my/cycle-theme)
#+end_src

*** evil
#+begin_src emacs-lisp :tangle yes
(setq +evil-want-o/O-to-continue-comments nil)
#+end_src
*** Haskell
#+begin_src emacs-lisp :tangle yes
(map! :map haskell-mode-map
      :localleader
      (:prefix-map ("l" . "lookup on hoogle")
       :desc "lookup word" "w" #'hoogle-word-under-caret
       :desc "lookup region" "r" #'hoogle-region)
      (:prefix-map ("r" . "run haskell process")
       :desc "load file" "r" #'haskell-process-load-file)
      :desc "switch to repl" "s" #'haskell-interactive-switch)

(map! :map interactive-haskell-mode-map
      :localleader
      :desc "switch to code" "s" #'haskell-interactive-switch-back)
#+end_src


.doom.d/config.el
;;; .doom.d/config.el -*- lexical-binding: t; -*-

;;; Theme and Fonts -------------------------------------
(setq doom-theme 'doom-palenight)

(setq doom-font (font-spec :family "Fira Code" :size 16)
      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 12)
      doom-big-font-mode nil
      doom-big-font-increment 2)

;; Bold keywords
(set-face-attribute 'font-lock-keyword-face nil :weight 'bold)

;; show unicode glyphs in different font
(set-fontset-font t 'unicode "all-the-icons" nil 'prepend)

;; unprettify prettified symbols when under cursor
(setq prettify-symbols-unprettify-at-point 'right-edge)

;; Ligatures in Fira Code
(global-fira-code-mode)
(setq fira-code-mode-disabled-ligatures '("x" "[]"))


;;; Keybindings ----------------------------------------
;; General
(define-key prog-mode-map (kbd "C-c f")
  'format-buffer)

(map! :leader
      :desc "Raise popup"
      "w m p" #'+popup/raise)

;; Haskell
(map! :map haskell-mode-map
      :localleader
      (:prefix-map ("l" . "lookup on hoogle")
       :desc "lookup word" "w" #'hoogle-word-under-caret
       :desc "lookup region" "r" #'hoogle-region)
      (:prefix-map ("r" . "run haskell process")
       :desc "load file" "r" #'haskell-process-load-file)
      :desc "switch to repl" "s" #'haskell-interactive-switch)

(map! :map interactive-haskell-mode-map
      :localleader
      :desc "switch to code" "s" #'haskell-interactive-switch-back)

;; org
(add-hook! org-mode
  (lambda ()
    (local-set-key (kbd "<C-return>") 'org-insert-heading)))

;;; -----------------------------------------------------

;;; evil ------------------------------------------------

;; o/O should not continue commented lines
(setq +evil-want-o/O-to-continue-comments nil)

;;; org-mode --------------------------------------------
;;Visual Stuff
(add-hook! org-mode (lambda ()
                      (org-bullets-mode t)
                      (yas-minor-mode t)))
;; Do not prettify symbols
(setq org-pretty-entities nil)

(setq org-ellipsis "⤵")

;;; -----------------------------------------------------

;;; ivy -------------------------------------------------
;; enable fuzzy matching in ivy
(setq ivy-re-builders-alist
      '((t . ivy--regex-plus)))
;;; -----------------------------------------------------

;;; treemacs

;; Do not show files ignored by git in treemacs
(with-eval-after-load 'treemacs
  (add-to-list 'treemacs-pre-file-insert-predicates #'treemacs-is-file-git-ignored?))
;;; -----------------------------------------------------

;;; company
(setq company-tooltip-idle-delay 0.2)
(setq company-idle-delay 0.2)
(setq company-show-numbers t)
;;; -----------------------------------------------------

;;; recentf

;; Do not show files in the .emacs.d directory in recent files
(with-eval-after-load 'recentf
  (add-to-list 'recentf-exclude
               (expand-file-name "~/.emacs.d//")))
;;; -----------------------------------------------------

;;; Haskell-Mode
(setq haskell-process-type 'stack-ghci
      haskell-process-path-stack "stack"
      haskell-process-args-stack-ghci '("--ghci-options=-ferror-spans"))

(setq haskell-interactive-popup-errors nil)

(defun hoogle-word-under-caret ()
  "Lookup the word under the caret on hoogle"
  (interactive)
  (save-excursion
    (forward-word)
    (let ((end (point)))
      (backward-word)
      (hoogle-region (point) end))))

(defun hoogle-region (start end)
  "Lookup the region between start and end on hoogle"
  (interactive "r")
  (haskell-hoogle (buffer-substring start end)))

;;; Utilities -------------------------------------------
(defun format-buffer ()
  "Format the whole buffer"
  (interactive)
  (set-face-attribute 'font-lock-keyword-face nil :weight 'bold)
  (delete-trailing-whitespace)
  (indent-region (point-min) (point-max) nil)
  (untabify (point-min) (point-max)))
;;; -----------------------------------------------------
